cpp/ql/src/semmle/code/cpp/Class.qll

```diff
@@ -614,6 +614,15 @@ class Class extends UserType {
     class_template_argument(underlyingElement(this), i, unresolveElement(result))
   }
 
+  /**
+   * Gets the `i`th template argument value used to instantiate this class from a
+   * class template. When called on a class template, this will return the
+   * `i`th template parameter value.
+   */
+  override Expr getTemplateArgumentValue(int i) {
+    class_template_argument_value(underlyingElement(this), i, unresolveElement(result))
+  }
```

Comment:
Would it be better to just generalise
```
Type getTemplateArgument(int i)
```
to
```
Element getTemplateArgument(int i)
```
(and similarly in the dbscheme)?

Reply:
I think the QL-level API you proposed there is more intuitive than what's proposed in this PR, and it would also cut down on the number of similar-looking predicates that are added. Unfortunately it's a breaking change to change the return type because it would no longer be possible to write `getTemplateArgument(0).getName()`. It might be a breaking change we're willing to make since I don't expect too many external queries to use this API at all.

We also currently get the declared type of a non-type template parameter extracted and accessible via `getTemplateArgument` (the `myIntAlias` in `template<myIntAlias i> ...`). If we want to keep that, we'd also need an accessor predicate for it. I'd call it the [_kind_](https://en.wikipedia.org/wiki/Kind_(type_theory)) of the template argument, but maybe there's a more appropriate C++-specific word for it.

Reply:
My initial design of the interface had the goal to not break anything, and to keep the changes small.  I agree its not necessarily intuitive, and therefore am happy to change it.

However, I think having a predicate which returns either a `Type` or `Expr` depending on the template parameter kind is not good either as this means that those who want to know the type of `Expr` need to use another predicate to get that type.

I think there should be two predicates:
 * One to get the type of a template argument (either explicitly specified by the user, or deduced by the compiler).  This has a value for every template argument.
 * One to get the value of the template argument.  Only has a value for non-type templates.

(This is actually the current interface - although it could have much better names).

Maybe a better approach would be something like the following (which I think is closer to how the C++ Standard views the world):

Consider the following:
```
template<typename T, T X> T foo() { T y = X; return y + y; } 
int x = foo<int, 10>(); 
```
Currently. the trap we produce for the instantiation of `foo<int, 10>` currently does not reference the type `T` anywhere - instead the return type and type of `y` are directly `int`.  Similarly, `X` is replaced with `10`.

Maybe instead we should expose `T` as a typedef and `X` as a variable declaration, and use those instead.  In the example above this would mean the return type of the instantiation of `foo()` is `T` which is a typedef for `int`.

Then in the template declaration we have `getTemplateParameter()` which return the names `T`, `X` (in this example) and provide a link to the types in the instantiation.  (This final bit isn't clearly thought out :-) ).

My quick scan of the C/C++ Front-End source suggests that it holds enough info in the IL for this to be workable.

This is definitely a breaking change, and much more complicated to implement, but would be the more natural view of the world.  However, does what we're currently doing provide enough for end users?

Reply:
I think the most natural interface would be, if `inst` is your `foo<int, 10>` instantiation,
```
inst.getTemplateArgumentKind(0) = type // currently not representable
inst.getTemplateArgument(0) = int
inst.getTemplate().getTemplateArgumentKind(0) = typename // currently not representable
inst.getTemplate().getTemplateArgument(0) = T

inst.getTemplateArgumentKind(0) = int
inst.getTemplateArgument(0) = 10
inst.getTemplate().getTemplateArgumentKind(1) = T
inst.getTemplate().getTemplateArgument(1) = X
```

Reply:
I like your idea of using typedefs and variables. With C++11 `template<...> using` declarations, I think this will even generalise to template template parameters. But for non-type template parameters, it's unclear to me how we'd map the variable to its value. If an instantiated template has variable `X` for a non-type template parameter, should the `Initializer` of that variable then be the expression that holds its value (`10`, in your example)? What if the same template instantiation is found in multiple source locations? One location might have `foo<int, 10>`, and another might have `foo<int, 5+5>`.

The interface currently implemented in this PR is not great, but I don't think it's an interface that will be used much, so I'm fine with merging it as it is as long as there is extensive QLDoc about all the surprising behaviour we've discussed.

Reply:
@igfoo I agree that would be an ideal interface (I think there's a typo: the last two `(0)` should be `(1)`). I don't know if the idea of a _kind system_ is natural to people with a background in C++ rather than type theory.

Is the distinction between `type` and `typename` important? Or can we get away with having a single kind system for template parameters and arguments?

Reply:
@igfoo: I think this is where we disagree.  I don't like the idea that there's a predicate whose result is either a `Type` or an `Expr` and if its an `Expr` you have to go looking elsewhere to find its type.  But I suppose that just becomes a wrapper predicate:
```
predicate getTemplateArgumentType(Declaration d, int i) {
  if d.getTemplateArgument(i).instanceOf(Type)
  then result = d.getTemplateArgument(i)
  else result = d.getTemplateArgumentKind(i)
}
```

I wouldn't change the proposed DB Schema however. As the type is always present and so should be in the `@template_argument()` tuple.  We then have a separate tuple for the value of non-type templates - as those are not present on all template arguments.

Reply:
@jbj: Given the front-end behaviour for `foo<int, 10>` and `foo<int, 5 + 5>` we would get just one instantiation with the expression `10` for X.  So you can't distinguish between the two.  This is also the case with my current patchset.

This makes sense from the FE's point-of-view - these are both the same template instantiation, and a back-end would only generate the code once.

Reply:
Yes

Reply:
igfoo:  Can you clarify something from your example above? 

You wrote:

> inst.getTemplateArgumentKind(0) = type // currently not representable
> inst.getTemplateArgument(0) = int
> inst.getTemplate().getTemplateArgumentKind(0) = typename // currently not representable
> inst.getTemplate().getTemplateArgument(0) = T
> 
> inst.getTemplateArgumentKind(0) = int
> inst.getTemplateArgument(0) = 10
> inst.getTemplate().getTemplateArgumentKind(1) = T
> inst.getTemplate().getTemplateArgument(1) = X

I think the last block should all refer to index 1, and so read:

> **inst.getTemplateArgumentKind(*1*) = int
> inst.getTemplateArgument(*1*) = 10
> inst.getTemplate().getTemplateArgumentKind(1) = T
> inst.getTemplate().getTemplateArgument(1) = X

Is that correct?

Reply:
Yes; I've edited it accordingly.

Reply:
Yes I did - updated :-).

I could imagine writing something like:

```template<class T, T::size_type initial_size> array```

where I would want to know the underlying type of `T::size_type` (4 or 8 bytes?).

This is where treating the parameters as typedefs and variable decls would be more natural because `initial_size` would be a variable of type `T::size_type`.

Reply:
Anyway where I think we've got to is the need for two predicates on Declaration:

1. `Element getTemplateArgument(int i)` which returns the user-supplied `i`-th template argument.
2. `Type getTemplateArgumentKind(int i) which returns the type of the `i`-th template argument if that is a non-type argument, and has no result otherwise.

Correct?

Reply:
And that the above is a breaking change - but we do not expect it to cause many people issues (if any).

Reply:
I've implemented the above change - although the return type of both getTemplateArgument and getTemplateArgumentKind is now Locatable as that is the common-parent of Type & Expr.